docs/source/notebooks/AutoEncoders.md

```diff
@@ -73,94 +73,83 @@ print(x_test.shape)
+class AutoEncoder(BaseWrapper, TransformerMixin):
+    """A class that enables transform and fit_transform.
     """
-    return keras.Model(fit_model.get_layer("input").input, fit_model.get_layer("encoded").output)
+    
+    def _keras_build_fn(self, encoding_dim: int, meta: Dict[str, Any]):
+        n_features_in = meta["n_features_in_"]
 
-def get_inverse_tf_model(fit_model: keras.Model, encoding_dim: int) -> keras.Model:
-    """Get an deencoder model.
+        encoder_input = keras.Input(shape=(n_features_in,))
+        encoder_output = keras.layers.Dense(encoding_dim, activation='relu')(encoder_input)
+        encoder_model = keras.Model(encoder_input, encoder_output)
 
-    We do this by extracting the deencoding layer from the fitted autoencoder model
-    and adding a new Keras input layer.
-    """
-    encoded_input = keras.Input(shape=(encoding_dim,))
-    output = fit_model.get_layer("output")(encoded_input)
-    return keras.Model(encoded_input, output)
-```
+        decoder_input = keras.Input(shape=(encoding_dim,))
+        decoder_output = keras.layers.Dense(n_features_in, activation='sigmoid', name="decoder")(decoder_input)
+        decoder_model = keras.Model(decoder_input, decoder_output)
+        
+        autoencoder_input = keras.Input(shape=(n_features_in,))
+        encoded_img = encoder_model(autoencoder_input)
+        reconstructed_img = decoder_model(encoded_img)
 
-Next we create a class that that will enable the `transform` and `fit_transform` methods, as well as integrating all three of our models into a single estimator.
+        autoencoder_model = keras.Model(autoencoder_input, reconstructed_img)
 
-```python
-from sklearn.base import TransformerMixin, clone
-from scikeras.wrappers import BaseWrapper
-
-
-class KerasTransformer(BaseWrapper, TransformerMixin):
-    """A class that enables transform and fit_transform.
-    """
-
-    def __init__(self, *args, tf_est: BaseWrapper = None, inv_tf_est: BaseWrapper = None, **kwargs) -> None:
-        super().__init__(*args, **kwargs)
-        self.tf_est = tf_est
-        self.inv_tf_est = inv_tf_est
+        self.encoder_model_ = BaseWrapper(encoder_model, predict__verbose=self.predict__verbose)
+        self.decoder_model_ = BaseWrapper(decoder_model, predict__verbose=self.predict__verbose)
 
+        return autoencoder_model
 
     def fit(self, X, sample_weight=None):
         super().fit(X=X, y=X, sample_weight=sample_weight)
-        self.tf_est_ = clone(self.tf_est)
-        self.inv_tf_est_ = clone(self.inv_tf_est)
-        self.tf_est_.set_params(fit_model=self.model_)
-        self.inv_tf_est_.set_params(fit_model=self.model_, encoding_dim=self.encoding_dim)
-        X = self.feature_encoder_.transform(X)
-        self.tf_est_.initialize(X=X)
-        X_tf = self.tf_est_.predict(X=X)
-        self.inv_tf_est_.initialize(X_tf)
+        X_tf = self.encoder_model_.initialize(X).predict(X)
+        self.decoder_model_.initialize(X_tf)
```

Comment:
Shouldn't this code go add to `super().initialize`, not `super().fit`?

Reply:
I'm pretty sure we want to be calling `super().fit`, but I may be misunderstanding the question.

Reply:
Shouldn't this code be used?

``` python
def initialize(self, X_train):
    super().initialize(X_train)
    X_tf = self.encoder_model_.initialize(X_train).predict(X_train)
    self.decoder_model_.initialize(X_tf)
    return self
```

Reply:
I see, you are right. In fact, the way it is currently written, the encoder is getting `X` before it is modified by `target_encoder_`. This could be fixed by calling `target_encoder_` from `AutoEncoder.fit`, but this would result in an extra `target_encoder_` call which is not good.

But I don't think doing it the way you propose is currently possible. This is because `fit` does not call `initialize` but rather `_initialize` because we want to get back `X` & `y` instead of `self. We would have to do:

```python
class AutoEncoder(BaseWrapper, TransformerMixin):

   def _initialize(self, X, y):
       X, y = super()._initialize(X, y)
       X_tf = self.encoder_model_.initialize(X).predict(X)
      self.decoder_model_.initialize(X_tf)
      return X, y
    
    def fit(...):
      return super().fit(...)  # i.e. we no longer modify fit!
```

Reply:
This could be solved if `initialize` retuned `X, y` instead of `self`, but IMO returning `self` is nicer.

Reply:
It could also be solved if `_initialize` were public, correct?

In https://github.com/adriangb/scikeras/pull/200#issuecomment-782310802's `fit` function, I'm having a hard time seeing why no calls to `encoder_model_.fit` or `decoder_model.fit_` are made.

Reply:
> It could also be solved if _initialize were public, correct?

We have a public `initialize`, it's the same as `_initialize` except for the return arguments (`_initialize` returns `X` & `y`, `initialize` returns `self`).

> I'm having a hard time seeing why no calls to encoder_model_.fit or decoder_model.fit_ are made

`encoder_model_` and `decoder_model_`  share layers (and their weights) with `model_`. When `model_` gets fitted, the weights in `encoder_model_` and `decoder_model_` also get updated. Then `encoder_model_` and `decoder_model_` are used as inference-only models in `transform` and `inverse_transform` respectively. That is, it is completely valid to call `predict` on `encoder_model_` and `decoder_model_` without ever calling `fit`.

Reply:
Regarding the use of `_initialize`, the way it stands now the `AutoEncoder` class is a subclass of `BaseWrapper` which is overriding the protected member method `BaseWrapper._initialize` which is totally fine, at least in other languages where these types of things are actually enforced. Some better documentation of `_initialize` might help understand why and how developers subclassing `BaseWrapper` might want to call it, but I think that is beyond the scope of this PR.